clippy_lints/src/trivial_default_constructed_types.rs

```diff
@@ -0,0 +1,113 @@
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &Expr<'tcx>) {
+        if !expr.span.from_expansion()
+            && let ExprKind::Call(call, _) = expr.kind
+            && is_trait_item(cx, call, sym::Default)
+        {
+            let ret_ty = cx
+                .typeck_results()
+                .expr_ty(call)
+                .fn_sig(cx.tcx)
+                .output()
+                .skip_binder()
+                .peel_refs();
+            if let Some(default) = default_value(cx, ret_ty) && !is_from_proc_macro(cx, expr) {
+                span_lint_and_sugg(
+                    cx,
+                    TRIVIAL_DEFAULT_CONSTRUCTED_TYPES,
+                    expr.span,
+                    "constructing a trivial type using `default`",
+                    "try",
+                    default.to_string(),
+                    Applicability::MachineApplicable,
+                );
+            } else if let ty::Tuple(fields) = ret_ty.kind()
+                && fields.len() <= 3
+                && let Some(fields_default) = fields.iter()
+                    .map(|field| default_value(cx, field))
+                    .collect::<Option<Vec<_>>>()
+                && !is_from_proc_macro(cx, expr)
+            {
+                let default = if let [default] = &*fields_default {
+                    // Needs trailing comma to be a single-element tuple
+                    format!("{default},")
+                } else {
+                    fields_default.join(", ")
+                };
+
+                span_lint_and_sugg(
+                    cx,
+                    TRIVIAL_DEFAULT_CONSTRUCTED_TYPES,
+                    expr.span,
+                    &format!(
+                        "constructing a {} using `default`",
+                        if fields.is_empty() { "unit" } else { "trivial tuple" },
+                    ),
+                    "try",
+                    format!("({default})"),
+                    Applicability::MachineApplicable,
+                );
+            } else if let ty::Array(ty, len) = ret_ty.kind()
+                && let Some(default) = default_value(cx, *ty)
+                && !is_from_proc_macro(cx, expr)
+            {
+                span_lint_and_sugg(
+                    cx,
+                    TRIVIAL_DEFAULT_CONSTRUCTED_TYPES,
+                    expr.span,
+                    "constructing a trivial array using `default`",
+                    "try",
+                    format!("[{default}; {len}]"),
+                    Applicability::MachineApplicable,
+                );
+            }
+        }
+    }
+}
+
+/// Gets the default value of `ty`.
+fn default_value(cx: &LateContext<'_>, ty: Ty<'_>) -> Option<Cow<'static, str>> {
+    match ty.kind() {
+        ty::Adt(def, substs) if let [subst] = substs.as_slice() => {
+            is_lang_item_or_ctor(cx, def.did(), LangItem::Option).then(|| format!("None::<{subst}>").into())
```

Comment:
Adding the type argument to `None` when it's not needed isn't really an improvement. `None::<u32>` is ok. `None::<Vec<String>>` is not great. `None::<Arc<Mutex<SomeStruct<String>>>>` is terrible. None of them are better than `Option::default()`.

Reply:
We would likely need to use the `QPath` instead of `Ty` to fix that, but this would ignore type aliases and the like

Reply:
I think the easiest way to fix this is to just make it `HasPlaceholders` in the `None` case, and use `::<_>`.

Reply:
You bring up a fun issue with type aliases. Numeric literals can't just use type suffixes if the required type is an alias since it could be configuration dependent.

With all the edge cases I don't think there's any way to avoid parsing the type part of the path. Only linting when either the type is missing (e.g. `Default::default()`), or the type is a single identifier matching the expected type's name (e.g. `Option::default()`) is fine. If somebody does something stupid like `type u32 = i32;` then the lint can not trigger. `type MyInt = i32;` is probably better not linted as well.

Reply:
That's a good point, so parsing the path is necessary then. I believe we can use `.into()` for type aliases in particular, or `<name>::from(x)` perhaps (which is usually what I use for creating type aliases). But we probably shouldn't lint them yeah

Reply:
I'm unfortunately not sure how this is supposed to both avoid type aliases and lint `Default::default`. We have to use a ty afaik in that case so we end up skipping type aliases. I have it working for everything else but it's basically useless for arrays, slices, tuples etc if `Default::default` isn't linted

Reply:
This is easier to do now with `expr_use_ctxt`. You'll have to add a function to get the defined type as a HIR type, but that's mostly copying the existing function. This won't work for anything defined in another crate unfortunately.

Reply:
Honestly I feel like at this point it'd just be easier to store where it's a type alias directly in the `Ty`'s info 😅 It can probably be changed upstream rather easily

I may look into that...

Reply:
You would still need to take type inference into account, so you would still have to look up the type as it was written either way.

Reply:
Indeed, but at least we'd be able to ignore type aliases more easily. Not a big deal for `u32::default` where `u32` is `type u32 = i32` but `Default::default` is currently pretty annoying

Reply:
I think, for now, I'll change this to lint type aliases for `Default::default` and ignoring them on `u32::default` using the `Res`. One day, once `lazy_type_alias` is stable (and hopefully that normalization is delayed until after typeck :sweat_smile:) then we can ignore them entirely.